### PR TITLE
fixes entry point bug reported on slack 

### PR DESCRIPTION
**Pull request by tjfulle**
See https://spackpm.slack.com/archives/C5W7NKZJT/p709823515239109 and #42370

### CODE REVIEW

**Comment by haampie** [comment]
I just opened https://github.com/spack/spack/pull/43079 ;p but feel free to take it over. I did not fix the test.

File: lib/spack/llnl/util/lang.py [hunk 1]
```diff
@@ -858,32 +856,17 @@ def get_entry_points(*, group: str):
     """
 
     try:
+        import importlib.metadata  # type: ignore  # novermin
```
Comment (haampie) [inline lib/spack/llnl/util/lang.py:857]: avoid nested try blocks for readable backtraces

File: lib/spack/spack/test/entry_points.py [hunk 2]
```diff
@@ -66,18 +66,8 @@ def entry_points(group=None):
...
-        except ImportError:
-            import importlib_metadata
-        monkeypatch.setattr(importlib_metadata, "entry_points", entry_points)
-    except ImportError:
-        try:
-            import pkg_resources  # type: ignore
-        except ImportError:
-            return
-        monkeypatch.setattr(pkg_resources, "iter_entry_points", entry_points)
+    import importlib.metadata # type: ignore # novermin
```
Comment (haampie) [inline lib/spack/spack/test/entry_points.py]: Better not to mock functions you don't own, mock `get_entry_points` instead.

Comment (tjfulle) [reply lib/spack/spack/test/entry_points.py]: That's fine with me - but then `get_entry_points` will be untested...

Comment (haampie) [reply lib/spack/spack/test/entry_points.py]: Yup, but it's run in various other places. Edit: ok, your new unit test is fine.

File: lib/spack/spack/test/entry_points.py [hunk 3]
```diff
@@ -95,7 +86,7 @@ def test_spack_entry_point_config(tmp_path, mock_entry_points):
 
 
 @pytest.mark.skipif(sys.version_info[:2] < (3, 8), reason="Python>=3.8 required")
```
Comment (haampie) [inline lib/spack/spack/test/entry_points.py]:
```suggestion
```

File: lib/spack/spack/test/entry_points.py [hunk 4]
```diff
@@ -64,24 +66,13 @@ def entry_points(group=None):
...
-    except ImportError:
-        try:
-            import pkg_resources  # type: ignore
-        except ImportError:
-            return
-        monkeypatch.setattr(pkg_resources, "iter_entry_points", entry_points)
+    monkeypatch.setattr(llnl.util.lang, "get_entry_points", entry_points)
 
 
 @pytest.mark.skipif(sys.version_info[:2] < (3, 8), reason="Python>=3.8 required")
```
Comment (haampie) [inline lib/spack/spack/test/entry_points.py]:
```suggestion
```

File: lib/spack/spack/test/entry_points.py [hunk 5]
```diff
@@ -110,3 +101,16 @@ def test_spack_entry_point_extension(tmp_path, mock_entry_points):
     assert os.path.samefile(root, my_ext)
     module = spack.extensions.get_module("spam")
     assert module is not None
+
+
+@pytest.mark.skipif(sys.version_info[:2] < (3, 8), reason="Python>=3.8 required")
+def test_llnl_util_lang_get_entry_points(tmp_path, monkeypatch):
+    import importlib.metadata  # type: ignore # novermin
+
+    monkeypatch.setattr(importlib.metadata, "entry_points", entry_points_factory(tmp_path))
```
Comment (haampie) [inline lib/spack/spack/test/entry_points.py:108]: fair

**Review by haampie — DISMISSED** [review]

File: lib/spack/llnl/util/lang.py [hunk 6]
```diff
@@ -847,26 +847,24 @@ def get_entry_points(*, group: str):
...
-            entry_points = importlib.metadata.entry_points(group=group)
-        except TypeError:
-            # Prior to Python 3.10, entry_points accepted no parameters and always
-            # returned a dictionary of entry points, keyed by group.  See
-            # https://docs.python.org/3/library/importlib.metadata.html#entry-points
-            entry_points = importlib.metadata.entry_points().get(group, [])
-        yield from entry_points
     except ImportError:
-        yield from ()
+        return []
```
Comment (haampie) [inline lib/spack/llnl/util/lang.py:859]: fwiw, could also make the return type list since that's a supertype.

**Review by haampie — APPROVED** [review]

**Pull request merged**